app/models/amazon_agent_manager.rb

```diff
@@ -0,0 +1,393 @@
+      }]
+    }
+
+    role = iam.create_role(
+      role_name: role_name,
+      assume_role_policy_document: policy_doc.to_json
+    )
+
+    # grant all priviledges
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonS3FullAccess'
+    })
+
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonEC2FullAccess'
+    })
+
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonSQSFullAccess'
+    })
+
+    role
+  end
+
+  def create_profile(profile_name = MIQ_SSA, role_name = MIQ_SSA)
+    ssa_profile = iam.instance_profile(profile_name)
+    ssa_profile = iam.create_instance_profile(instance_profile_name: profile_name) unless ssa_profile.exists?
+
+    find_or_create_role(role_name)
+    ssa_profile.add_role(role_name: role_name) if ssa_profile.roles.zero?
+
+    ssa_profile
+  end
+
+  # possible RHEL image name: values: [ "RHEL-7.3_HVM_GA*" ]
+  def get_agent_image_id(img = agent_ami)
+    imgs = ec2.client.describe_images(
+      filters: [
+        {
+          name: "name",
+          values: [img]
+        }
+      ]
+    ).images
+
+    imgs[0].image_id
+  end
+
+  def create_security_group(group_name = MIQ_SSA)
+    sgs = ec2.client.describe_security_groups(
+      filters: [
+        {
+          name: "group-name",
+          values: [group_name]
+        }
+      ]
+    ).security_groups
+    return sgs[0].group_id unless sgs.empty?
+
+    # create security group if not exist
+    security_group = ec2.create_security_group({
+      group_name: group_name,
+      description: 'Security group for MIQ SSA Agent',
+      vpc_id: ec2.client.describe_vpcs.vpcs[0].vpc_id
+    })
+
+    security_group.authorize_ingress({
+      ip_permissions: [{
+        ip_protocol: 'tcp',
+        from_port: 22,
+        to_port: 22,
+        ip_ranges: [{
+          cidr_ip: '0.0.0.0/0'
+        }]
+      }]
+    })
+
+    security_group.authorize_ingress({
+      ip_permissions: [{
+        ip_protocol: 'tcp',
+        from_port: 80,
+        to_port: 80,
+        ip_ranges: [{
+          cidr_ip: '0.0.0.0/0'
+        }]
+      }]
+    })
+
+    security_group.authorize_ingress({
+      ip_permissions: [{
+        ip_protocol: 'tcp',
+        from_port: 443,
+        to_port: 443,
+        ip_ranges: [{
+          cidr_ip: '0.0.0.0/0'
+        }]
+      }]
+    })
+
+    security_group.group_id
+  end
+
+  # Get Key Pair for SSH. Create a new one if not exists.
+  def get_key_pair(pair_name = MIQ_SSA)
+    kps = Authentication.where(name: pair_name)
+    kps.each do |kp|
+      return kp if kp.resource_id == @ems.id
+    end
+
+    ManageIQ::Providers::CloudManager::AuthKeyPair.create_key_pair(@ems.id,
+      { :key_name => pair_name })
+  end
+
+  def create_pem_file(pair_name = MIQ_SSA)
+    kp = get_key_pair(pair_name)
+    pem_file_name = pair_name + ".pem"
+    File.open(pem_file_name, 'w') { |f| f.write(kp.auth_key) }
+    File.chmod(0400, pem_file_name)
+    pem_file_name
```

Comment:
Why are we writing the pem file to disk?  I could see if the aws_sdk needs it that way, but then I'd expect that we should treat that pemfile like a tempfile and blow it away/recreate on every invocation so that it doesn't leak.

Reply:
This method isn't called, so we don't write pem file to disk. It's here for future debugging use: retrieve pem from DB, write it to disk and use it to ssh to the agent. After problem is solved, the pem file should be removed.

Reply:
If a method isn't called, we should't keep it, IMO.  If your purpose is diagnostic, then it's not SSA specific, so you should add it to the KeyPair class directly, and then add some specs for it.